File: 0496.Next Greater Element I/stack.cpp
class Solution {
public:
    vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {
        stack<int> stk;
        unordered_map<int, int> m;
        vector<int> ret;
        
        for(const int& n: nums2){
            while(stk.size() > 0 && stk.top() < n){
                m[stk.top()] = n;
                stk.pop();
            }
            stk.push(n);
        }
        
        for(const int& n: nums1){
            if(m.find(n) != m.end())
                ret.push_back(m[n]);
            else
                ret.push_back(-1);
        }
        
        return ret;
    }
};

File: 0496.Next Greater Element I/stack.java
class Solution {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        Stack<Integer> monotonicStack = new Stack<>();
        Map<Integer, Integer> map = new HashMap<>();
        
        int len1 = nums1.length, len2 = nums2.length;
         int[] ret = new int[len1];
        for(int i: nums2){
            while(monotonicStack.size() > 0 && monotonicStack.peek() < i){
                map.put(monotonicStack.pop(), i);
            }
            monotonicStack.push(i);
        }
        
        for(int i = 0; i < len1; ++i){
            if(map.containsKey(nums1[i])){
                ret[i] = map.get(nums1[i]);
            }else{
                ret[i] = -1;
            }
        }
        return ret;
    }
}